File: Python Fundamentals/Mid Exam problems/man_o_war.py
pirate_ship = list(map(int, input().split('>')))
warship = list(map(int, input().split('>')))
max_hp = int(input())

while True:
    command = input()
    if command == "Retire":
        break
    command_list = command.split()

    if command_list[0] == 'Fire':
        if 0 <= int(command_list[1]) <= len(warship) - 1:
            warship[int(command_list[1])] -= int(command_list[2])
            if warship[int(command_list[1])] <= 0:
                print("You won! The enemy ship has sunken.")
                exit()

    elif command_list[0] == 'Defend':
        if 0 <= int(command_list[1]) <= len(pirate_ship) - 1 and 0 <= int(command_list[2]) <= len(pirate_ship) - 1:
            for attacked_section in range(int(command_list[1]), int(command_list[2]) + 1):
                pirate_ship[attacked_section] -= int(command_list[3])
                if pirate_ship[attacked_section] <= 0:
                    print("You lost! The pirate ship has sunken.")
                    exit()

    elif command_list[0] == 'Repair':
        if 0 <= int(command_list[1]) <= len(pirate_ship) - 1:
            pirate_ship[int(command_list[1])] += int(command_list[2])
            if pirate_ship[int(command_list[1])] > max_hp:
                pirate_ship[int(command_list[1])] = max_hp

    elif command_list[0] == 'Status':
        print(f"{len([x for x in pirate_ship if x < 0.20 * max_hp])} sections need repair.")

print(f'Pirate ship status: {sum(pirate_ship)}')
print(f'Warship status: {sum(warship)}')

File: JavaScript Fundamentals/Mid Exams/Mid Exam 6/manoWar.js
function solve(arr) {
    let pirateship = arr.shift().split('>').map(x => Number(x));
    let warship = arr.shift().split('>').map(x => Number(x));
    let maxHp = Number(arr.shift());
    let stalemate = true;

    for (let el of arr) {
        if (el == 'Retire') break;
        let tokens = el.split(' ');
        if (tokens[0] == 'Fire') {
            let index = Number(tokens[1]);
            let damage = Number(tokens[2]);
            if (index >= 0 && index < warship.length) {
                warship[index] -= damage;
                if (warship[index] <= 0) {
                    console.log("You won! The enemy ship has sunken.");
                    stalemate = false;
                    break;
                }
            }
        } else if (tokens[0] == 'Defend') {
            let startIndex = Number(tokens[1]);
            let endIndex = Number(tokens[2]);
            let damage = Number(tokens[3]);
            let areIndicesValid = true;
            for (let checkIndex = startIndex; checkIndex <= endIndex; checkIndex++) {
                if (!(checkIndex >= 0 && checkIndex < pirateship.length)) {
                    areIndicesValid = false;
                    break;
                }
            }

            if (areIndicesValid) {
                for (let attackIndex = startIndex; attackIndex <= endIndex; attackIndex++) {
                    if (attackIndex >= 0 && attackIndex < pirateship.length) {
                        pirateship[attackIndex] -= damage;
                        if (pirateship[attackIndex] <= 0) {
                            console.log("You lost! The pirate ship has sunken.");
                            stalemate = false;
                            break;
                        }
                    }
                }
            }

        } else if (tokens[0] == 'Repair') {
            let index = Number(tokens[1]);
            let health = Number(tokens[2]);
            if (index >= 0 && index < pirateship.length) {
                pirateship[index] += health;
                if (pirateship[index] > maxHp) pirateship[index] = maxHp;
            }


        } else if (tokens[0] == 'Status') {
            let count = 0;
            for (let section of pirateship) {
                if (section < 0.20 * maxHp) count++;
            }
            console.log(`${count} sections need repair.`);
        }
    }
    if (stalemate) {
        let pirateSum = 0;
        let warSum = 0;
        for (i of pirateship) pirateSum += i;
        for (j of warship) warSum += j;

        console.log(`Pirate ship status: ${pirateSum}`);
        console.log(`Warship status: ${warSum}`);
    }
}
